### ISSUE TITLE
Scheduler, StatefulSets, and External Controllers

### ISSUE THREAD

**Issue by foxish** [kubernetes/kubernetes#39687]
The implementation of equivalence classes currently [ignores StatefulSets](https://github.com/kubernetes/kubernetes/blob/9705bb728e768be02bf5b32aa14dec059f9269ed/plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go#L252-L260). The reason seems to be https://github.com/kubernetes/kubernetes/issues/32024#issuecomment-244685920. I think the reasoning no longer holds given the goals and the current implementation of the StatefulSet. It does not create pods with different resource shapes and this can be removed safely.

- Equivalence class Issue: https://github.com/kubernetes/kubernetes/issues/32024
- Equivalence class Implementation PR: https://github.com/kubernetes/kubernetes/pull/31605

From the standpoint of fork-ability of StatefulSets and similar controllers, we do not want the scheduler to have any special casing of the StatefulSet kind.  However, it exposes a different concern, one of external controllers that do manage non-uniform pods, which would need a way to tell the scheduler that they cannot each be treated as equivalent.

cc @davidopp @kubernetes/sig-scheduling-misc @kubernetes/sig-apps-misc

**Comment by timothysc** [comment]
@jayunit100 assigning

**Comment by jayunit100** [comment]
I assume the problem of "treating pods as equivalent" is more related to the concern of rescheduling then anything else, right?  because the use of an equivalence class to simply predicate match is inoccuous wether or not a pod is stateful... 

Assuming the above, then to uplevel, this issue is a good example of the general problem of scheduler logic conflicting with external controllers... and i guess there are two categories here, where the intersection of *what the scheduler wants to do* can either :

(1) *subvert* other features (like statefullness) or 

(2) be *subverted*  by other issues (i.e. for example https://github.com/kubernetes/kubernetes/issues/39687).  

Solutions to (1) are likely to conflict with solutions to (2).  For example, aggressive rescheduling gaurantees scheduler policy is adhered to and fixed, but could break any number of features like StatefulSets which we otherwise care about. 

Now, if we assume other pod mutating processes are imperfect ( I agree on 4301 w/ @timothysc that rescheduling is a better solution then forcing a dependency on the scheduler for everyone that wants to do a one off scheduling event), then we must take rescheduling as a necessary feature to fix asymmetrical affinity or other things that can be caused by downscaling...

... if thats the case we have solved problem (2), but possibly inflamed problem (1) ... so a simple rectification is then to give pods a way to communicate, generically, to the scheduler that they are immune to rescheduling ...

**Comment by smarterclayton** [comment]
Rescheduling needs to have a controller independent way to trigger safe
transformation of pod sets.  That sounds like disruption budget.

I might be missing the thread of your argument, but rescheduler can break
stateful and stateless applications equally - a 2 node stateless HA router
controlling ingress to your platform is broken if rescheduler knocks them
off at the same time, just like an active/passive MySQL instance.

Since we decided StatefulSets aren't wildly different, I agree with Anirudh
we should remove this check.

On Jan 10, 2017, at 6:17 PM, jay vyas <notifications@github.com> wrote:

I assume the problem of "treating pods as equivalent" is more related to
the concern of rescheduling then anything else, right? because the use of
an equivalence class to simply predicate match is inoccuous wether or not a
pod is stateful...

Assuming the above, then to uplevel, this issue is a good example of the
general problem of scheduler logic conflicting with external controllers...
and i guess there are two categories here, where the intersection of *what
the scheduler wants to do* can either :

(1) *subvert* other features (like statefullness) or

(2) be *subverted* by other issues (i.e. for example #39687
<https://github.com/kubernetes/kubernetes/issues/39687>).

Solutions to (1) are likely to conflict with solutions to (2). For example,
aggressive rescheduling gaurantees scheduler policy is adhered to and
fixed, but could break any number of features like StatefulSets which we
otherwise care about.

Now, if we assume other pod mutating processes are imperfect ( I agree on
4301 w/ @timothysc <https://github.com/timothysc> that rescheduling is a
better solution then forcing a dependency on the scheduler for everyone
that wants to do a one off scheduling event), then we must take
rescheduling as a necessary feature to fix asymmetrical affinity or other
things that can be caused by downscaling...

... if thats the case we have solved problem (2), but possibly inflamed
problem (1) ... so a simple rectification is then to give pods a way to
communicate, generically, to the scheduler that they are immune to
rescheduling ...

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/39687#issuecomment-271729361>,
or mute the thread

**Comment by jayunit100** [comment]
I think we all agree on the change, the remaining question the last one posed in the original issue : is external controllers tell the scheduler not to touch certain pods?

  I guess a Disruption budget of 0 would prevent any destabilization ; but that would disable any evictions entirely.... I assume folks would **want** to support eviction while still telling the rescheduler not to rebalance automagically .

**Comment by davidopp** [comment]
I didn't read the whole thread but I think you may be missing some of the context. (Note that I didn't review the implementation, so I don't know exactly what was implemented, but I know what the goal is.)

Equivalence class is a way to identify a set of pods as identical with respect to scheduling requirements, so that you only need to run the predicate/priority functions across all nodes for one pod in the equivalence class, and then you can reuse the result for the remaining pods in the equivalence class. (It's a little more complicated than that, since obviously the state of some nodes change as stuff from that equivalence class or others are scheduled, but you get the idea.)

Most Kubernetes controllers create sets where all the pods are identical, so there's no need to do anything fancy like compute a hash of the pod template to identify the equivalent pods -- you know all the pods of a ReplicaSet, Job, etc. are identical. StatefulSet is the one exception to this.

If you don't want to hard-code the names of controllers to distinguish the ones that generate identical replicas vs. non-identical replicas, you have to do something else to identify sets of equivalent pods.

cc/ @wojtek-t

**Comment by foxish** [comment]
I agree. I vote for deleting the special casing of the StatefulSet from the scheduler for now. 

As for the more general issue, we would either want a way to use an annotation or field to specify if a controller is managing homogeneous or heterogeneous pods, or for the scheduler to figure that out from inspection of resource requirements of individual pods. But I agree with @davidopp that we can discuss that when we get there.

**Comment by davidopp** [comment]
Yeah, the "normal" way to identify equivalent pods is to hash (as a single unit) the fields of the pod template that the scheduler cares about, and build a map from hash to set of pods (that are all equivalent).

**Comment by k82cn** [comment]
@davidopp, when we're talking about scheduler, suppose it's default scheduler and for **all kind of workload**. If we're going to support all kind of workload in Kubernetes, I agree with @jayunit100 to add some flags which let "rescheduler doesnt touch certain pods": for some "computing" workload, e.g. MPI, it'll lost internal result which makes those "computing" workload to re-run. But for others kind of workload, I agree that re-scheduler is helpful for cluster utilization.

**Comment by davidopp** [comment]
I think we should move the discussion of how to ensure rescheduler operates "safely" and how it would prioritize/deprioritize which pod(s) to kill when it has a choice, to the design issue for rescheduler (#12140).  But in general I think it comes down to mechanisms like PDB and priority, and appropriate quota on those so that there isn't an "arms race" where everyone puts "max disruption == 0" and "priority == infinite" on all of their pods.

**Comment by k82cn** [comment]
sure, will continue the discussion at #12140 

> "max disruption == 0" and "priority == infinite" on all of their pods

Great 👍

**Comment by wojtek-t** [comment]
Yes - we shouldn't mix rescheduler here. The real rescheduler doesn't still exists and is not even designed. So all discussions about rescheduler should be moved there. 

And i agree that removing StatefulSet from the check was a good thing to do for now.
